### PR TITLE
Redesign postbox handling

### PR DESCRIPTION
**Pull request by JoosuaKoskinen**
#### What are the main changes in this PR?

Redesigns the way Seravo Plugin tools, postboxes and AJAX calls are created and managed. There's no high level postbox templates, AJAX templates or postbox components yet.

Main features:
- Allows more reusable code and unified design
- Takes automatically care of caching in most cases
- Already has a couple of automatic AJAX responses built-in
- Is able of catching exceptions and showing prettier errors 

All old postboxes will work normally and they will be upgraded one by one once there's enough templates and helpers for them.

##### Why are we doing this? Any context or related work?

Getting Seravo Plugin up-to-date.

#### Where should a reviewer start?

Proper review will be hard but small logic flaws are no problem as this will adapt anyways as old code is upgraded and we know better what we need. The important part is that old code isn't affected.

Please tell me if there's a better design for something and please be picky about comments etc, let's do this the right way.

#### Manual testing steps?

If you want, you can try making a test postbox, here's one example:
https://github.com/Seravo/seravo-plugin/commit/722ef3bdae589d52256b95b73d239677919727c1

NEW EXAMPLE: https://github.com/Seravo/seravo-plugin/tree/example/postbox

### CODE REVIEW

File: lib/postbox/postbox.php [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+     * @param int   $cache_time Seconds to cache data for (default is 0).
+     */
+    public function set_data_func( $data_func, $cache_time = 0 ) {
+      $this->data_func = $data_func;
+      $this->data_cache_time = $cache_time;
+    }
+
+    /**
+     * @param array<string, mixed>|Requirements $requirements Requirements for this postbox.
+     *                                                        Array in "[Requirements::*] => mixed" format.
```
Comment (Moppa5) [inline lib/postbox/postbox.php:179]: Weird format 🤔, phpcs / rector ?

**Review by Moppa5 — COMMENTED** [review]
I wanted to give some comments even though this is marked as WIP. In general I think that you have done great job with this new design and refactoring 👍🏼. I don't really have any major comments regarding the design. However, see some minor "nagging" and comments / questions below.

File: js/postbox/seravo-ajax.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+'use strict'
+
```
[the remark this answers is not in the document]
Comment (JoosuaKoskinen) [reply js/postbox/seravo-ajax.js:2]: Yes, that's right. This file wasn't supposed to be committed yet and will be finished today.

File: lib/postbox/ajax.php [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+     * @param string $section Unique section inside the postbox.
+     */
+    public function __construct( $section ) {
+      $this->section = $section;
+    }
+
+    /**
+     * @param string $id    Unique id/slug of the postbox.
+     * @param string $nonce Name of WordPress nonce for the page.
+     */
```
[the remark this answers is not in the document]
Comment (JoosuaKoskinen) [reply lib/postbox/ajax.php:63]: Would be nice yes. I don't think that very obvious functions need to be commented but this commit still needs a lot of commenting.

File: lib/postbox/component.php [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+      $component = new Component();
+      $component->set_content($html);
+
+      return $component;
+    }
+
+
+    /**#########################
+     * ### MID LEVEL HELPERS ###
+     * #########################*/
```
[the remark this answers is not in the document]
Comment (JoosuaKoskinen) [reply lib/postbox/component.php:120]: I think component.php is the right place for them but they could be separated from the Component -class.

File: lib/postbox/postbox.php (hunk 1 above)
Comment (JoosuaKoskinen) [reply lib/postbox/postbox.php:179]: No, this weird format is on me :) Will work on it

File: lib/postbox/component.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (Moppa5) [reply lib/postbox/component.php:120]: If you think it's the right place, let's go with that then 👍🏼

File: lib/postbox/ajax.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (Moppa5) [reply lib/postbox/ajax.php:63]: Whereas I would like to document even the obvious functions etc. for the sake of consistency, I understand your point. Maybe some functions like `isValid()` may not need commenting as they are clear by the context.

[the remark this answers is not in the document]
Comment (JoosuaKoskinen) [reply lib/postbox/ajax.php]: Ended up commenting everything, please let me know if I missed some.

File: lib/postbox/postbox.php (hunk 1 above)
Comment (JoosuaKoskinen) [reply lib/postbox/postbox.php]: Fixed

File: lib/postbox/component.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (JoosuaKoskinen) [reply lib/postbox/component.php]: I ended up creating template.php where all the functions for getting pre-made components are placed.

**Comment by JoosuaKoskinen** [comment]
New example commit: https://github.com/Seravo/seravo-plugin/tree/example/postbox

File: js/postbox/seravo-ajax.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (JoosuaKoskinen) [reply js/postbox/seravo-ajax.js]: Removed the JS for now, will add back later.

**Review by Moppa5 — APPROVED** [review]
I think we are ready to merge this. I couldn't find any major issues by statically analyzing the code. As mentioned we can work and find possible issues better when we start making something concrete.

**Pull request merged**